### ISSUE TITLE
Language in DDI exporter

### ISSUE THREAD

**Issue by lubitchv** [IQSS/dataverse#9266]
Dataverse has a metadatafield "Language". But it seems there is no way to determine the language of dataset in DDI export. It seems that the only field that has language attribute is subject keyword, but it is always English, disregarding the dataset language.

**What steps does it take to reproduce the issue?**
Create a dataset with metadatafield "Language" selected as non-English, for example French. Publish dataset. Export DDI through UI "Metadata->DDI" or Native API.
Field stdyDscr/stdyInfo/subject/keyword" will be always with attribute "xml:lang="en" disregarding the language of the study and the controlled vocabulary will be always in English. There is no other way to determine the language of the dataset.   

* When does this issue occur?
During DDI export.

* To whom does it occur (all users, curators, superusers)?
All users that use DDI export.

* What did you expect to happen?
""xml:lang" attribute and controlled vocabulary to depend on language of the dataset.

**Comment by qqmyers** [comment]
I think what you want is the metadatalanguage - see https://guides.dataverse.org/en/latest/installation/config.html?highlight=metadatalanguage#allowing-the-language-used-for-dataset-metadata-to-be-specified . 

My understanding is that the Language metadata field is meant to allow you to specify the language(s) used in the data files. The :MetadataLanguages setting/mechanism was added to allow admins/collection managers to decide which languages they will accept metadata in and to then use the value selected by the person creating that dataset to show in the edit/display of the datasetpage and metadata export docs.

**Comment by lubitchv** [comment]
Thanks @qqmyers. It helps, although the problem is that we have datasets in two languages in the same dataverse collection. We want to determine the language of a dataset from DDI export. We need it for our migration project.

**Comment by qqmyers** [comment]
I think it supports that. For new datasets, you can set a collection to allow French or English and the user creating the dataset picks which one at creation time. There isn't any way to allow both languages/duplicate fields for each language. For existing datasets, I think you could set the metadatalanguage if you know it (possible only through db edits, I don't recall if there's an superuser api for that.)

**Comment by lubitchv** [comment]
Thanks @qqmyers, I see now that xml:lang attribute appears in export codeBook field if user select language. I also noticed that in subject keyword it appears in both languages:
`<keyword xml:lang="en">Agricultural Sciences</keyword>
<keyword xml:lang="fr">Sciences de l'agriculture</keyword> `
The question I have how to do it through api DDI import, Does Metadata Language can be setup through xml:lang of codeBook?

**Comment by qqmyers** [comment]
I think so - see https://github.com/IQSS/dataverse/blob/ac1454b50ee892bdb410d4dd96a56cba6883394d/src/main/java/edu/harvard/iq/dataverse/api/imports/ImportDDIServiceBean.java#L186 . There also may be examples/discussion in the guides. I know we added some for some methods, but I'm not sure if DDI was one.

**Comment by lubitchv** [comment]
Hmm.. when I added 
``curl http://localhost:8080/api/admin/settings/:MetadataLanguages -X PUT -d '[{"locale":"en","title":"English"},{"locale":"fr","title":"Français"}]'`` 
I was not able to run import DDI API at all.
``curl -H X-Dataverse-key:XXX -X POST http://localhost:8080/api/dataverses/:root/datasets/:importddi --upload-file ddi.xml``
I get an error ``{"status":"ERROR","code":400,"message":"Bad Request. The API request cannot be completed with the parameters supplied. Please check your code for typos, or consult our API guide at http://guides.dataverse.org.","requestUrl":"http://localhost:8080/api/v1/dataverses/:root/datasets/:importddi","requestMethod":"POST"}``

Just before running curl command with MetadataLanguages I was able to import the same xml file. It maybe something that codebook attributes that are not compatible now? Since before it  was not important and that xml has non-empty codebook attributes.

**Comment by lubitchv** [comment]
Yes, in our xml files from that we want to import we have xml-lang attribute instead of xml:lang attribute. In this case it fails to load all together, but after changing it into xml:lang it is imported ok. 
It means that xml:lang is necessary attribute without which import DDI will always fail.
So it will resolve our issue of determining language from export DDI.

Although, for dataverses that do not have determined language but dataset citation does have language field set, the export would not determine language. What do you guys think about such case?

**Comment by qqmyers** [comment]
Not sure I fully understand your question. I think not using anything that maps to the Language field in the Citation block to determine the metadata language is 'by design' since that appear to relate to the file language(s). Not allowing you to send the metadatalanguage in via the metadatalanguage key in json, which is what the xml:lang in ddi maps to, when metadatalanguage is not configured was also ~by design (to avoid providing a way to do that via API when you couldn't through the UI, to avoid having datasets advertised with metadata in languages the admins don't want to support, etc.) I think the hope by keeping the :MetadataLanguages setting separate from the UI :Languages setting, that admins could potentially have a long list if they wanted to support the metadata language in exports.

That said, proposals to change those defaults are fine - we just need to make sure it wouldn't break an existing installation/provide some backward compatible option if it would. Also, w.r.t. to using the Language field, I think we're close to having a PR to allow alternate exporters, which could allow you to decide to replace the current DDI exporter(s) with custom ones that, for example, assume the metadata is always the first entry for the Language field, etc., i.e. implementing something that might not be universal but would work for your instance.

**Comment by lubitchv** [comment]
@qqmyers I talked to the team about MetadataLanguage feature. The objection was that it forces dataset creator to pick up the language of the dataset (during creation). It is outside of Metadata section. This language may also be potentially different from language in citation block. This leads to user's confusion and potential inconsistencies.

But we still would like to determine the language of the dataset from export DDI. We would prefer if it would be possible to do from the language in metadata citation block. Codebook 2.5 does not have language field in citation, it exists only in DDI-Lifecycle 3.2. So we suggest in case :MetadataLanguages is not determined inside settings table, to write xml:lang attribute in codeBook field that is taken from citation language of a dataset in Dataverse. This is for export DDI.
For import DDI, we suggest to make this attribute optional and if it is determined, to set the citation language to it. It would be very useful since at the moment there is no way to setup citation language in dataverse through DDI import and this is a gap between DDI and dataverse representation. 

This would be compatible with existing features.

**Comment by qqmyers** [comment]
This might be something to discuss with @jggautier since he maintains crosswalks to other vocabularies.

For backward compatibility, I think use cases like harvesting from a system using the :MetadataLanguages capability would have to be considered. For a dataset with a French description of files in English (where the Language field in the citation block is en), it sounds like your change would import this into a Dataverse without :MetadataLanguages with only the Language field (en) which you would then interpret as the language of the metadata? I'm not sure what other areas where using that field according to the current guidelines ("A language that the Dataset's files is written in" - the field appears to allow multiples as well) would cause problems if you add a second interpretation (how would you handle multiples?). 

That's one reason why the initial design ended up with a separate setting - to avoid such misinterpretations/complexity. The other was that the metadatalanguage really is a different kind of metadata - it is metadata about the metadata, it might be something admins want to limit to assure the metadata is readable by their audience, etc. If/when we allow metadata in more than one language at once (there's at least an issue about it, I don't know of any work), maintaining a link between the languages field and the metadata language would get even harder.

FWIW, if it helps, I think it would be possible to make the metadatalanguage choice changeable after creation. That was done partly to limit scope and partly to avoid having someone flip it without editing all the fields. Making edit-after-creation an option would provide backward compatibility.

**Comment by jggautier** [comment]
Hi all. Could we have a Zoom call about this? I'm having a hard time following. There's lots of chances for me to misinterpret what's been written so far. Maybe a conversation in real time would help.

**Comment by cmbz** [comment]
@lubitchv are you still interested in this feature? Please contact @jggautier
